test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamples.cs

```diff
@@ -665,11 +643,36 @@ private void MixMatch(string dataPath)
             var model = staticFinalPipe.Fit(data);
 
             // And here is how we could've stayed in the dynamic pipeline and train that way.
-            dynamicPipe = dynamicPipe.Append(new KeyToValueEstimator(env, "PredictedLabel"));
+            dynamicPipe = dynamicPipe.Append(new KeyToValueEstimator(mlContext, "PredictedLabel"));
             var dynamicModel = dynamicPipe.Fit(data.AsDynamic);
 
             // Now 'dynamicModel', and 'model.AsDynamic' are equivalent.
-            var rs = model.Transform(data).GetColumn(x => x).ToArray();
+        }
+
+        [Fact]
+        public void ReadData()
+        {
+            ReadDataDynamic(GetDataPath("generated_regression_dataset.csv"));
+        }
+
+        private void ReadDataDynamic(string dataPath)
+        {
+            // Create a new context for ML.NET operations. It can be used for exception tracking and logging, 
+            // as a catalog of available operations and as the source of randomness.
+            var mlContext = new MLContext();
+
+            // Create the reader: define the data columns and where to find them in the text file.
+            var reader = mlContext.Data.TextReader(new[] {
+	                // We read the first 10 values as a single float vector.
+                    new TextLoader.Column("FeatureVector", DataKind.R4, new[] {new TextLoader.Range(0, 9)}),
+                    // Separately, read the target variable.
+                    new TextLoader.Column("Target", DataKind.R4, 10)
+                },
+                // Default separator is tab, but we need a comma.
+                s => s.Separator = ",");
```

Comment:
(This is probably not exactly related to `MLContext`, but since you wrote this new test, I'll ask the question):

Do we like having `s =>` here? Why isn't the separator just an argument in the `TextReader()` method?

Reply:
The pattern we want to have is that we add the most important options as ctor options, and the rest 'advanced' inside the delegate.
The other pattern is that we want the extensions to replicate the signatures of constructors, to avoid any mapping logic in strange places. 
Because of that, I'd rather change the signature of `TextLoader` ctor to add 'hasHeader' and 'separator', but we will need to do it in a separate PR.

Also 'separator' should not really be a string :)

---
In reply to: [226106086](https://github.com/dotnet/machinelearning/pull/1252#discussion_r226106086) [](ancestors = 226106086)